see: https://github.com/ayufan-rock64/linux-build/issues/112#issuecomment-572153549
and: https://github.com/ayufan-rock64/linux-build/files/4036318/0001-usb-dwc3-disable-park-mode.patch.zip
and: https://lkml.org/lkml/2019/10/14/496

From 174496dd7371da29fca5c4edf9b09c0f3a7c3c07 Mon Sep 17 00:00:00 2001
From: Li Jun <jun.li@nxp.com>
Date: Tue, 29 Oct 2019 10:25:35 +0800
Subject: [PATCH] usb: dwc3: disable park mode

Park mode advantages is minimal, synopsys has a known issue on
this, so disable it.

Signed-off-by: Li Jun <jun.li@nxp.com>
---
 drivers/usb/dwc3/core.c | 12 ++++++++++++
 drivers/usb/dwc3/core.h |  1 +
 2 files changed, 13 insertions(+)

diff --git a/drivers/usb/dwc3/core.c b/drivers/usb/dwc3/core.c
index 95b79c7..adc8ece 100644
--- a/drivers/usb/dwc3/core.c
+++ b/drivers/usb/dwc3/core.c
@@ -895,6 +895,18 @@ static int dwc3_core_init(struct dwc3 *dwc)
 		if (dwc->dis_tx_ipgap_linecheck_quirk)
 			reg |= DWC3_GUCTL1_TX_IPGAP_LINECHECK_DIS;
 
+		/*
+		 * Synopsys STAR:
+		 * USB3.0 HC died when read and write at the same
+		 * time with park mode.
+		 * It has advantage only a single async EP is
+		 * active, which is meaningless for application;
+		 * So disable park mode and synopsys will change
+		 * the default value of park mode to be disabled
+		 * in next release.
+		 */
+		reg |= DWC3_GUCTL1_PARKMODE_DISABLE;
+
 		dwc3_writel(dwc->regs, DWC3_GUCTL1, reg);
 	}
 
diff --git a/drivers/usb/dwc3/core.h b/drivers/usb/dwc3/core.h
index f7e0891..6ddeb8e 100644
--- a/drivers/usb/dwc3/core.h
+++ b/drivers/usb/dwc3/core.h
@@ -210,6 +210,7 @@
 /* Global User Control 1 Register */
 #define DWC3_GUCTL1_TX_IPGAP_LINECHECK_DIS	BIT(28)
 #define DWC3_GUCTL1_DEV_L1_EXIT_BY_HW	BIT(24)
+#define DWC3_GUCTL1_PARKMODE_DISABLE	BIT(17) | BIT(16) | BIT(15)
 
 /* Global USB2 PHY Configuration Register */
 #define DWC3_GUSB2PHYCFG_PHYSOFTRST	BIT(31)
-- 
2.7.4

